### PR TITLE
[ENGINE] Fail engine if Lucene commit fails

### PR DESCRIPTION
**Pull request by s1monw**
This is similar to refresh, if we fail to commit the data we have to fail the
engine since in-ram data is likely discarded. Yet, it's still in translog and might
be recoverable when the node is restarted but we have to treat the engine as failed.

### CODE REVIEW

**Comment by bleskes** [comment]
LGTM. Should we also put this into 1.4?

**Comment by s1monw** [comment]
> LGTM. Should we also put this into 1.4?

we might - you can see this as a bug though.... I think I agree....

File: src/main/java/org/elasticsearch/index/engine/internal/InternalEngine.java [hunk 1]
```diff
@@ -743,7 +754,7 @@ public void flush(FlushType type, boolean force, boolean waitIfOngoing) throws E
                         { // commit and close the current writer - we write the current tanslog ID just in case
                             final long translogId = translog.currentId();
                             indexWriter.setCommitData(Collections.singletonMap(Translog.TRANSLOG_ID_KEY, Long.toString(translogId)));
-                            indexWriter.commit();
+                            commitIndexWriter(indexWriter);
```
Comment (bleskes) [inline src/main/java/org/elasticsearch/index/engine/internal/InternalEngine.java]: I think it's a problem to call fail engine here as we're holding the write lock. failEngine has a code block that acquires the read lock: 

```
} finally {
                closedOrFailed = true;
                try (InternalLock _ = readLock.acquire()) {
                    // we take the readlock here to ensure nobody replaces this IW concurrently.
                    if (indexWriter != null) {
                        indexWriter.rollback();
                    }
                } catch (Throwable t) {
                    logger.warn("Rolling back indexwriter on engine failure failed", t);
                    // to be on the safe side we just rollback the IW
                }
            }
```

Maybe we should reset indexWriter to null and then skip the read lock in failEngine if that's the case. Not sure.

Comment (s1monw) [inline src/main/java/org/elasticsearch/index/engine/internal/InternalEngine.java]: downgrading a writeLock is fine - the problems is different that some caller holds the read lock and we fail the engine wich in-turn closes it (and acquires a write lock)

**Comment by s1monw** [comment]
closing this - this will be fixed in https://github.com/elasticsearch/elasticsearch/issues/9616

**Pull request closed without merging**